Add configurable log message formatting

FrostyFlippper opened the pull request:
Introduces a config file for customizing log message arrangement. Logger and DefaultLogger now support setting a config file, and log messages are formatted based on its contents.

FrostyFlippper committed:
Add configurable log message formatting

Introduces a config file for customizing log message arrangement. Logger and DefaultLogger now support setting a config file, and log messages are formatted based on its contents.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+
+        //arrange using config file
+        try {
+            if(!Files.exists(this.configFile)){
```
```suggestion
            if(Files.notExists(this.configFile)){
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+        String loggingLevelName = level.name();
+        String loggerName = this.name;
+
+        //arrange using config file
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+        //arrange using config file
+        try {
+            if(!Files.exists(this.configFile)){
+                //default config
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+            }
+
+            String logMessageArrangement = Files.readString(this.configFile);
+            Map<String, Object> m = new HashMap<>();
```
`m` is a terrible variable name

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+        long seconds = localTime.getSecond();
+        long milliseconds = localTime.getNano() / 1_000_000;
+        long nanoseconds = localTime.getNano();
+
+        String threadName = Thread.currentThread().getName();
+        String loggingLevelName = level.name();
+        String loggerName = this.name;
+
+        //arrange using config file
+        try {
+            if(!Files.exists(this.configFile)){
+                //default config
+                Files.writeString(this.configFile, "{hours}:{minutes}:{seconds} [{threadName}] {loggingLevelName} {loggerName} - {message}", StandardOpenOption.CREATE_NEW);
+            }
+
+            String logMessageArrangement = Files.readString(this.configFile);
+            Map<String, Object> m = new HashMap<>();
+
+
+            m.put("{hours}", hours);
+            m.put("{minutes}", minutes);
+            m.put("{seconds}", seconds);
+            m.put("{milliseconds}", milliseconds);
+            m.put("{nanoseconds}", nanoseconds);
+
+            m.put("{threadName}", threadName);
+            m.put("{loggingLevelName}", loggingLevelName);
+            m.put("{loggerName}", loggerName);
+            m.put("{message}", message);
+
+
+            for(Map.Entry<String, Object> entry :  m.entrySet()){
+                logMessageArrangement = logMessageArrangement.replace(entry.getKey(), String.valueOf(entry.getValue()));
+            }
+
+            return logMessageArrangement;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
All the variables can be inlined. Saves 9 lines of code, since the variables arent used anywhere else.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+
+
+            for(Map.Entry<String, Object> entry :  m.entrySet()){
+                logMessageArrangement = logMessageArrangement.replace(entry.getKey(), String.valueOf(entry.getValue()));
```
```suggestion
                logMessageArrangement = logMessageArrangement.replace(entry.getKey(), Objects.toString(entry.getValue()));
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -138,6 +141,52 @@ public void log(String message, LoggingLevel level, Object... objects) {
+            }
+
+            return logMessageArrangement;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
```suggestion
        } catch (IOException exception) {
            throw new RuntimeException("Something went wrong parsing the log config file!", exception);
        }
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -207,17 +255,18 @@ private void beginWriteScheduling() {
         private boolean isCompressionEnabled = true;
-        private long logFrequency = TimeUnit.SECONDS.toMillis(1); // Default to 1 second
-        private long deletionFrequency = TimeUnit.DAYS.toMillis(1); // Default to 1 day
+        private long logFrequency = java.util.concurrent.TimeUnit.SECONDS.toMillis(1); // Default to 1 second
+        private long deletionFrequency = java.util.concurrent.TimeUnit.DAYS.toMillis(1); // Default to 1 day
```
```suggestion
        private long logFrequency = TimeUnit.SECONDS.toMillis(1); // Default to 1 second
        private long deletionFrequency = TimeUnit.DAYS.toMillis(1); // Default to 1 day
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/Logger.java:
```diff
@@ -152,6 +152,8 @@ default void debug(String logMessage, Object... objects) {
      */
     void addLogFile(Path file);
 
+    void setConfigFile(Path configFile);
```
Javadocs please

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -207,17 +255,18 @@ private void beginWriteScheduling() {
     public static class Builder {
+        //default builder attributes
         private final String name;
         private DateTimeFormatter loggingDateFormat = DateTimeFormatter.ofPattern("HH:mm:ss");
```
This won't be needed anymore, so you might as well remove it and everywhere it's used.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -207,17 +255,18 @@ private void beginWriteScheduling() {
      * Builder for creating a DefaultLogger instance.
      */
     public static class Builder {
+        //default builder attributes
```
```suggestion
```

DaRealTurtyWurty commented on src/test/java/Test.java:
```diff
@@ -17,6 +17,7 @@ public static void main(String[] args) throws InterruptedException {
                 .addLogFile("yes/no.log")
                 .isCompressionEnabled(true)
                 .loggingLevel(LoggingLevel.INFO)
+                .configFile(Path.of("src\\test\\config.txt"))
```
```suggestion
                .configFile(Path.of("src/test/config.txt"))
```

DaRealTurtyWurty reviewed (changes requested):
In addition to the requested changes, a config should be a JSON file, not a random text file, so that we can define additional data in it in the future, beyond just the log format.

FrostyFlippper committed:
switched to using notExists as it is cleaner

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
removed unecessary comment

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
removed unecessary comment

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
fixed a bug that could cause a null pointer exception

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
remove unecessary code

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
improved error messaging for io exception

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
Add config file support and logging layout customization

Introduces a JSON-based config file (config.json) for logger configuration, including customizable logging layouts. Updates DefaultLogger to read and write this config, removes unused date format logic, and adds Gson as a dependency. Adjusts builder defaults and cleans up related code and tests.
Updated the gitignore to include the logger config file as this is created automatically

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -75,9 +74,12 @@ public class DefaultLogger implements Logger {
-    DefaultLogger(String name, DateTimeFormatter loggingDateFormat, DateTimeFormatter logDateFormat) {
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+
+    private String loggingLayout = "{hours}:{minutes}:{seconds} [{threadName}] {loggingLevelName} {loggerName} - {message}";
```
Should be defined in the builder like everything else

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -87,6 +89,17 @@ public void init() {
+            JsonObject json = GSON.fromJson(Files.readString(this.configFile), JsonObject.class);
+            fromJson(json);
+        } catch (IOException exception) {
+            throw new RuntimeException(exception);
```
```suggestion
            throw new RuntimeException("An error has occurred loading the config file", exception);
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -134,57 +147,57 @@ public void log(String message, LoggingLevel level, Object... objects) {
+                JsonPrimitive loggingLayoutPrimitive = loggingLayoutElement.getAsJsonPrimitive();
+                if (loggingLayoutPrimitive.isString()) {
+                    this.loggingLayout = loggingLayoutElement.getAsString();
+                    System.out.println(this.loggingLayout);
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -196,7 +209,7 @@ public void close() {
             loggingMessages.clear();
             for (Path logFile : this.filesToLogTo) {
-                Files.writeString(logFile, logText + "\n", StandardOpenOption.APPEND, StandardOpenOption.CREATE);
+                Files.writeString(logFile, logText + "\n", StandardOpenOption.APPEND, java.nio.file.StandardOpenOption.CREATE);
```
```suggestion
                Files.writeString(logFile, logText + "\n", StandardOpenOption.APPEND, StandardOpenOption.CREATE);
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -242,7 +255,7 @@ private void beginWriteScheduling() {
                 this.loggingMessages.removeAll(messageCache);
                 for (Path logFile : this.filesToLogTo) {
-                    Files.writeString(logFile, logText + "\n", StandardOpenOption.APPEND, StandardOpenOption.CREATE);
+                    Files.writeString(logFile, logText + "\n", StandardOpenOption.APPEND, java.nio.file.StandardOpenOption.CREATE);
```
```suggestion
                    Files.writeString(logFile, logText + "\n", StandardOpenOption.APPEND, StandardOpenOption.CREATE);
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/logger/impl/DefaultLogger.java:
```diff
@@ -255,18 +268,16 @@ private void beginWriteScheduling() {
         private Path logDirectory = Path.of("logs");
         private final List<Path> filesToLogTo = new ArrayList<>();
         private boolean isCompressionEnabled = true;
         private long logFrequency = java.util.concurrent.TimeUnit.SECONDS.toMillis(1); // Default to 1 second
```
```suggestion
        private long logFrequency = TimeUnit.SECONDS.toMillis(1); // Default to 1 second
```

FrostyFlippper committed:
Added more detailed exception

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
removed debug print

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
remove unecessary inlined import

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
removed unecessary inline import

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
removed unecessary inline import

Co-authored-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

FrostyFlippper committed:
Add logging layout getter and setter to Logger interface

Introduced getLoggingLayout and setLoggingLayout methods to the Logger interface and implemented them in DefaultLogger. Also updated the DefaultLogger.Builder to support configuring the logging layout.

FrostyFlippper committed:
Merge branch 'master' of https://github.com/FrostyFlippper/RailroadLogger
